Allow config system customization

Pacman99 opened the pull request:
This allows users to pass the config system they want a profile built with and the output it should go in.
So you could now do:
```
configProfiles.NixOS = {
  builder = inputs.darwin.lib.darwinSystem;
  output = "darwinConfigurations";
}
```
Also added `defaultConfigBuilder` and `defaultConfigOutput` to make this a bit easier for those that only use one system. And renamed `nixosProfiles` to `configProfiles` with the proper deprecation line. But I actually think this should be named `hosts`, and then we can go `defaultHostOutput`. Or maybe configs could work.

This will prevent flake-utils-plus making an opinion on what config system, but nixos support is probably a bit better. And this should allow for darwin and home manager support. This could also work for robotnix,  but that would require a bit more rework to `configurationBuilder`

Pacman99 commented on systemFlake.nix:
```diff
@@ -5,14 +5,19 @@
+, nixosConfigurations ? { }
+  # nixosProfiles should be deprecated since profiles aren't nixos specific
+, nixosProfiles ? { }
+, configProfiles ? nixosProfiles
+, defaultConfigOutput ? "nixosConfigurations"
+, defaultConfigBuilder ? null
```
I think these should be renamed to hosts, but I'll let you decide first @gytis-ivaskevicius.

Pacman99 commented on systemFlake.nix:
```diff
@@ -48,34 +53,42 @@ let
+  configurationBuilder = hostname: profile: (
+    let
+      selectedNixpkgs = getNixpkgs profile;
+      defaultBuilder =
+        if defaultConfigBuilder == null
+        then inputs."${channelNameFromProfile profile}".lib.nixosSystem
+        else defaultConfigBuilder;
```
keep the nice nixos support, while still not sacrificing convenience for non-nixos.

Pacman99 commented on systemFlake.nix:
```diff
@@ -48,34 +53,42 @@ let
     }
   );
+
+  mapAttrs' = f: set:
+    builtins.listToAttrs (map (attr: f attr set.${attr}) (builtins.attrNames set));
```
This is a copy from upstream, I don't know if you want to put this somewhere else, but I'm fairly certain its necessary to get the `output` option for profiles to work.

gytis-ivaskevicius replied:
Maybe we should create a separate file with all the copy-pastes from nixpkgs? It could be just straight up added to flake lib argumentnt

Pacman99 replied:
@gytis-ivaskevicius what about putting it directly in `flake.nix`? since in staging systemflake now has access to flake-utils-plus lib.

gytis-ivaskevicius replied:
I was thinking about just dropping an import somewhere in here https://github.com/gytis-ivaskevicius/flake-utils-plus/blob/master/flake.nix#L26-L27 but adding it flake.nix directly isn't a bad idea, tho probably it does not scale well in the long run

Pacman99 replied:
Wait isn't the first link also `flake.nix`, and thats what I was talking about.

gytis-ivaskevicius replied:
All i mean is seperation of sorts like 
```nix
      flake-utils.lib
      // (import ./copy-pasta.nix {} )
      // rec {
      ```

Pacman99 replied:
ahh got it, I'll probably just add to flake.nix for now to avoid too many unrelated changes in this PR. But refactoring should definiitely happen

Pacman99 commented on systemFlake.nix:
```diff
@@ -106,9 +119,5 @@ otherArguments
   // optional defaultAppBuilder { defaultApp = defaultAppBuilder pkgs; }
   // optional devShellBuilder { devShell = devShellBuilder pkgs; }
   // optional checksBuilder { checks = checksBuilder pkgs; }
-)
-
-  // {
-  nixosConfigurations = nixosConfigurations // (builtins.mapAttrs nixosConfigurationBuilder nixosProfiles);
-}
+) // (mapAttrs' configurationBuilder configProfiles)
```
Not necessarily against the `nixosConfigurations //` part, but I'm not sure the best way to integrate with this.

Pacman99 replied:
considering https://github.com/gytis-ivaskevicius/flake-utils-plus/pull/15#issuecomment-814995499, `nixosConfiguations` is not going to work anyway. I would say just drop the escape hatch entirely, flake-utils-plus is general enough, where its pretty unecessary.

Pacman99 reviewed (commented):
Just a couple things to consider

Pacman99 commented on systemFlake.nix:
```diff
@@ -48,34 +53,42 @@ let
+        inherit (selectedNixpkgs) system;
+        modules = [
+          ({ pkgs, lib, ... }: {
+            networking.hostName = hostname;
+
+            nixpkgs = {
+              inherit (selectedNixpkgs) overlays config system;
+            };
+
+            system.configurationRevision = lib.mkIf (self ? rev) self.rev;
+            nix.package = lib.mkDefault pkgs.nixUnstable;
```
I think these work on darwin, but would deny home-manager support. I have a method to only set options if they exist, but it adds a bit of complexity:
```nix
{
type = "merge";
contents = [
  (optional (options ? nixpkgs) { nixpkgs = ... })
];
}
```

Pacman99 replied:
This would make this PR **actually** config system agnostic.

Pacman99 replied:
And then to ensure pkgs always gets passed at the very least it could be:
`if (options ? pkgs) { nixpkgs = ... } else { _module.args.pkgs = ... }`

Pacman99 commented on systemFlake.nix:
```diff
@@ -48,34 +53,42 @@ let
+        if defaultConfigBuilder == null
+        then inputs."${channelNameFromProfile profile}".lib.nixosSystem
+        else defaultConfigBuilder;
+      builder = profile.builder or defaultBuilder;
```
This is where the config system agnostic comes in. You can pass `darwinSystem` or `nixosSystem` or even home-manager configuration. The idea is just a function that builds the configuration.

blaggacao commented on systemFlake.nix:
```diff
@@ -106,9 +119,6 @@ otherArguments
-  nixosConfigurations = nixosConfigurations // (builtins.mapAttrs nixosConfigurationBuilder nixosProfiles);
-}
+) 
+  // (mapAttrs' configurationBuilder configProfiles)
```
```suggestion
# produces attrset in the shape of
# { nixosConfigurations = {}; darwinConfigurations = {};  ... } 
# according to profile.output or defaultConfigOutput
// (mapAttrs' configurationBuilder configProfiles)
```

blaggacao commented on systemFlake.nix:
```diff
@@ -5,14 +5,19 @@
+  # nixosProfiles should be deprecated since profiles aren't nixos specific
+, nixosProfiles ? { }
+, configProfiles ? nixosProfiles
+, defaultConfigOutput ? "nixosConfigurations"
```
```suggestion
, defaultConfigType ? "nixosConfigurations"
```

blaggacao replied:
The whole `ouput` string is actually duplicating information that we should the builder to produce by itself.

I cannot think of a case where there is not a 1:1 mapping between builder and it's output key.

So this input can be spared from the public facing api.

Pacman99 replied:
While that may be true, using the method I said, make the output of the builder '{ nixosConfigurations = }` is much harder to implement and adds complexity for the user trying to use it.
Its much easier to use to go
```nix
defaultConfigBuilder = inputs.darwin.lib.darwinSystem;
defaultConfigOutput = "darwinConfigurations;
```
And reading someone's flake.nix is much easier with this method.

Pacman99 replied:
```nix
defaultConfigBuilder = args: { darwinConfigurations = inputs.darwin.lib.darwinSystem };
```
Is very complex for the user, for us, and the reader of flake.nix.

Pacman99 replied:
Keeping output vs builder separation but dropped defaults, prefer #19.

blaggacao commented on systemFlake.nix:
```diff
@@ -5,14 +5,19 @@
+, nixosProfiles ? { }
+, configProfiles ? nixosProfiles
+, defaultConfigOutput ? "nixosConfigurations"
+, defaultConfigBuilder ? null
```
This line is a bit treacherous, it implies that there is _no_ `defaultConfigBuilder`

Pacman99 replied:
I think `null` and possibly a comment then dealing with the default later is much cleaner. Trying to make a default here is going to get unnecessary complex.

We could just add a comment, that `null` is not actually the default.

Pacman99 replied:
dropped these lines now defaults are set in code. Prefer #19 for defaults as not to increase flake-utils-plus api surface.

blaggacao commented on systemFlake.nix:
```diff
@@ -5,14 +5,19 @@
 , sharedModules ? [ ]
 , sharedOverlays ? [ ]
 
+, nixosConfigurations ? { }
```
```suggestion
```

This PR assumes resolution of #14 (see end of file)

Pacman99 replied:
You can read my comment there for reasoning, but I would say drop `nixosConfigurations` entirely. It goes against the idea that flake-utils-plus is a tool.

blaggacao commented on systemFlake.nix:
```diff
@@ -106,9 +114,8 @@ otherArguments
+) 
+# produces attrset in the shape of
+# { nixosConfigurations = {}; darwinConfigurations = {};  ... } 
+# according to profile.output or defaultConfigOutput
```
```suggestion
# according to profile.output or the default `nixosConfigurations`
```

blaggacao reviewed (approved):
LGTM, factoring out the way defaults are handled to #19 (for discussion) was an excellent idea! Thanks!

gytis-ivaskevicius commented on systemFlake.nix:
```diff
@@ -5,14 +5,17 @@
+, nixosConfigurations ? { }
+  # nixosProfiles should be deprecated since profiles aren't nixos specific
+, nixosProfiles ? { }
+, configProfiles ? nixosProfiles
```
Not a big fan of the name, maybe `hosts` or something like that should be considered.

Pacman99 replied:
I like hosts, but I'll probably submit the name change to a separate PR, as to discuss each change on its own.

gytis-ivaskevicius reviewed (approved):
I'd like to see some sort of example flake for all the changes that are being done. It does not need to be a good example or anything, just basic definitions are fine.

Pacman99 commented:
So I just switched to host in the top-level api and moved `mapAttrs'` to flake.nix as to keep this PR focused on its goal. I'll add an example next, and this should be ready to go.

Pacman99 commented on systemFlake.nix:
```diff
@@ -26,19 +27,21 @@
+    , system ? defaultHostAttrs.system or defaultSystem
+    , extraArgs ? defaultHostAttrs.extraArgs or { }
+    , output ? defaultHostAttrs.output or "nixosConfigurations"
+    , builder ? defaultHostAttrs.builder or channels.${channelName}.input.lib.nixosSystem
     , ...
-    }: defaultHostAttrs
-      // { 
-        inherit channelName system; 
-        modules = sharedModules ++ modules;
-        extraArgs = sharedExtraArgs // extraArgs;
-      };
+    }:
+    {
+      inherit channelName system output builder;
+      modules = sharedModules ++ modules;
+      extraArgs = sharedExtraArgs // extraArgs;
+    };
```
I realized defaults have to be set like this, otherwise `defaultHostAttrs` has no effect. Its a bit more verbose, but I think it looks pretty good.

blaggacao replied:
I don't agree with the semantics here:

`default*` and `shared*` ought to be equivalent alternatives.

Note, that the word "default" does not mandate "override" for mergables (`[]` & `{}`).

Pacman99 commented:
Added an un-tested nix-darwin example. Home-manager is possible, but requires writing a custom builder for it. 
I think this is ready, outside of the weird `nixosHosts` naming, which I would prefer to fix separately.

Pacman99 commented on systemFlake.nix:
```diff
@@ -5,16 +5,17 @@
-, sharedModules ? [ ]
 , sharedOverlays ? [ ]
 
+, nixosConfigurations ? { } # deprecate soon, no longer used or works
```
This is my solution to the "escape hatch" issue

blaggacao reviewed (changes requested):
Two things, as far as I could currently see

blaggacao commented on flake.nix:
```diff
@@ -33,6 +32,8 @@
           })
           paths;
 
+        mapAttrs' = f: set:
+          builtins.listToAttrs (map (attr: f attr set.${attr}) (builtins.attrNames set));
```
Nothing should be part of the public `fup` api, unless it adds serious `fup`-native value.

blaggacao replied:
Insofar, I guess I disagree with https://github.com/gytis-ivaskevicius/flake-utils-plus/pull/17#discussion_r609937344

Pacman99 replied:
So I just added mapAttrs' to systemFlake. Its either we export it or keep it in the file. I see no reason to use mapAttrs' outside of this use, so I think its ok to keep it in systemFlake.

blaggacao commented:
I once more conclude that our best option is to deprecate the top level api items that this can replace, for the following reasons:

- Don't risks to introduce subtle api differences ("default" vs "shared" vs "override")
- We avoid having to define and document precedence of competing declarations )of otherwise aliased options). Who has precedence? `defaultSystem` or `defaultHostsAttrs.system`? (To be clear: the point here is I do _not_ want to have to answer that question to keep the API reasonable).

Pacman99 commented:
Just cleaned up, I added the nixosHosts -> hosts rename here, since this is the PR where that becomes true. But overall I just followed the new host args merging code to include output and builder.

blaggacao commented on examples/somewhat-realistic-flake.nix:
```diff
@@ -73,7 +89,7 @@
 
-      # Shared modules/configurations between `nixosHosts`
+      # Shared modules/configurations between `hosts`
       defaultHostsAttrs = {
```
```suggestion
      hostsDefaults = {
```

blaggacao commented on examples/somewhat-realistic-flake.nix:
```diff
@@ -56,6 +57,21 @@
+        # This host uses `channels.unstable.{input,overlaysBuilder,config,patches}` attributes instead of `channels.nixpkgs.<...>`
+        channelName = "unstable";
+
+        # Host specific configuration. Same as `sharedModules`
```
```suggestion
        # Host specific configuration.
```

blaggacao commented:
Kudos for this great improvement! It is really a very well targeted generalization.
